{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "TabError",
     "evalue": "inconsistent use of tabs and spaces in indentation (<ipython-input-3-0451f7352b9c>, line 73)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-0451f7352b9c>\"\u001b[0;36m, line \u001b[0;32m73\u001b[0m\n\u001b[0;31m    if row['start'] > upstream_nucl:\u001b[0m\n\u001b[0m                                    ^\u001b[0m\n\u001b[0;31mTabError\u001b[0m\u001b[0;31m:\u001b[0m inconsistent use of tabs and spaces in indentation\n"
     ]
    }
   ],
   "source": [
    "#! /usr/bin/env python\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "import re\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "def get_degenes(filepath, gene_id, threshold, threshold_col_id):\n",
    "    genes = pd.read_csv(filepath)\n",
    "    genes = genes.dropna()   \n",
    "    if genes[gene_id].dtypes == float:\n",
    "        genes = genes.astype({gene_id:int})\n",
    "        genes = genes.astype({gene_id:str})\n",
    "        pass\n",
    "    elif genes[gene_id].dtypes == int:\n",
    "        genes = genes.astype({gene_id:str})\n",
    "    else:\n",
    "        print(\"gene names are strings, great!\")\n",
    "    \n",
    "    DEgenes = genes.loc[(genes[threshold_col_id] >= threshold) | (genes[threshold_col_id] <= -threshold)]\n",
    "    DEgenes = DEgenes[[gene_id]]\n",
    "    \n",
    "    return DEgenes\n",
    "\n",
    "def get_features(gff, gene_id, feature, search_gff, attribute, coord):\n",
    "    col_names = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    mygff = pd.read_csv(gff, sep='\\t', comment='#', low_memory=False, header=None, names=col_names)\n",
    "    CDS = mygff[mygff.type == feature]\n",
    "    CDS = CDS.copy()\n",
    "\n",
    "    RE_GENE_NAME = re.compile(r'({})(?P<gene_id>[0-9]+)'.format(search_gff))\n",
    "\n",
    "    def extract_gene_name(attributes_str):\n",
    "        res = RE_GENE_NAME.search(attributes_str)\n",
    "        return res.group('gene_id')\n",
    "    CDS[gene_id] = CDS.attributes.apply(extract_gene_name)\n",
    "    \n",
    "    RE_DESC = re.compile(r'({})(?P<attribute>.+?);'.format(attribute))\n",
    "    def extract_description(attributes_str):\n",
    "        res = RE_DESC.search(attributes_str)\n",
    "        if res is None:\n",
    "            return ''\n",
    "        else:\n",
    "            return res.group('attribute')\n",
    "    CDS['attribute'] = CDS.attributes.apply(extract_description)\n",
    "\n",
    "    CDS.drop('attributes', axis=1, inplace=True)\n",
    "    \n",
    "    if coord == 'all':\n",
    "        CDS_start_points = CDS\n",
    "    elif coord == 'min':\n",
    "        CDS_start_points = (CDS.groupby(['seqid', 'ENTREZID', 'attribute', 'strand'], as_index=False)['start'].min())\n",
    "    elif coord == 'max':\n",
    "        CDS_start_points = (CDS.groupby(['seqid', 'ENTREZID', 'attribute', 'strand'], as_index=False)['start'].max())\n",
    "    elif coord == 'median':\n",
    "        CDS_start_points = (CDS.groupby(['seqid', 'ENTREZID', 'attribute', 'strand'], as_index=False)['start'].median())\n",
    "        CDS_start_points = CDS_start_points.astype({'start':int})\n",
    "    else:\n",
    "        print('Non valid argument given to extract gene coordinates for start position')\n",
    "      \n",
    "    return CDS_start_points\n",
    "\n",
    "def create_background_fasta(CDS_start_points, gene_id, genome, background_outfile, upstream_nucl):\n",
    "    genome = Fasta(genome)\n",
    "    CDS_random = CDS_start_points.sample(500)\n",
    "    outfile = open(background_outfile, \"w\")   \n",
    "    back_list = []\n",
    "    back_dict= {}\n",
    "    \n",
    "    for index, row in CDS_random.iterrows():\n",
    "        genes = row[gene_id]\n",
    "        if row['start'] > upstream_nucl:\n",
    "            sequences = genome[row['seqid']][row['start'] - upstream_nucl:row['start'] + 5]\n",
    "        else:\n",
    "            sequences = genome[row['seqid']][row['start'] - row['start']:row['start']]\n",
    "        back_dict[genes] = sequences\n",
    "        back_list.append(back_dict)\n",
    "        back_dict = {}\n",
    "    \n",
    "    for d in back_list:\n",
    "        for key, value in d.items():\n",
    "            outfile.write(\">\" + key + \" \" + value.fancy_name + \"\\n\" + value.seq + \"\\n\")\n",
    "    \n",
    "    outfile.close() \n",
    "\n",
    "def create_target_fasta(DEgenes, CDS_start_points, gene_id, genome, target_outfile, upstream_nucl):\n",
    "    genome = Fasta(genome)\n",
    "    outfile = open(target_outfile, \"w\")\n",
    "    newdf = pd.merge(DEgenes, CDS_start_points)\n",
    "    seq_list = []\n",
    "    seq_dict= {}\n",
    "\n",
    "    for index, row in newdf.iterrows():\n",
    "        genes = row[gene_id]\n",
    "        if row['start'] > upstream_nucl:\n",
    "            sequences = genome[row['seqid']][row['start'] - upstream_nucl:row['start'] + 5]\n",
    "        else:\n",
    "            sequences = genome[row['seqid']][row['start'] - row['start']:row['start']]\n",
    "        seq_dict[genes] = sequences\n",
    "        seq_list.append(seq_dict)\n",
    "        seq_dict = {}\n",
    "\n",
    "    for d in seq_list:\n",
    "        for key, value in d.items():\n",
    "            outfile.write(\">\" + key + \" \" + value.fancy_name + \"\\n\" + value.seq + \"\\n\")\n",
    "    \n",
    "    outfile.close() \n",
    "\n",
    "def main():\n",
    "    \n",
    "    parser=argparse.ArgumentParser(description=\"get target and background sequences to use in oppossum\")\n",
    "    parser.add_argument(\"-filepath\", dest=\"filepath\", help=\"path to .csv file with differentially expressed genes\", type=str, required=True)\n",
    "    parser.add_argument(\"-gff\", dest=\"gff\", help=\"gff or path to gff file\", type=str, required=True)\n",
    "    parser.add_argument(\"-genome\", dest=\"genome\", help=\"genome fasta file\", type=str, required=True)\n",
    "    parser.add_argument(\"-target_out\", dest=\"target_outfile\", help=\"filename or path to output target sequences\", default=\"target_sequence.txt\", type=str)\n",
    "    parser.add_argument(\"-background_out\", dest= \"background_outfile\", help=\"filename or path to output background sequences\", type=str, default = \"background_sequence.txt\")\n",
    "    parser.add_argument(\"-g\",\"--gene_id\", dest =\"gene_id\", help=\"header of column with gene ids\", type=str, default=\"ENTREZID\")\n",
    "    parser.add_argument(\"-ti\",\"--threshold_id\", dest = \"threshold_col_id\", help=\"header of threshold column\", type=str, default = \"logFC\") # give default parameters\n",
    "    parser.add_argument(\"-th\",\"--threshold\", dest = \"threshold\", help=\"threshold number\", type=int, default = 2) # give default parameters \n",
    "    parser.add_argument(\"-sg\", \"--search_gff\", dest = \"search_gff\", help=\"search gff for gene id using a string\", type=str, default = \"GeneID:\")\n",
    "    parser.add_argument(\"-cd\", \"--feature\", dest = \"feature\", help=\"feature to extract from gff\", type=str, default = \"CDS\")\n",
    "    parser.add_argument(\"-at\", \"--attribute\", dest = \"attribute\", help=\"attribute to extract from gff\", type=str, default = \"product=\")\n",
    "    parser.add_argument(\"-ps\", \"--coordinate_position\", dest = \"coord\", help=\"start coordinate position to extract from gff\", type=str, choices=['all','min','max','median'], default = \"median\")\n",
    "    parser.add_argument(\"-u\", \"--upstream_nucl\", dest = \"upstream_nucl\", help=\"number of nucleotides to extract from start point of CDS\", type=int, default=5000)\n",
    "    args = parser.parse_args()\n",
    "    \n",
    "    DEgenes1 = get_degenes(filepath=args.filepath, gene_id=args.gene_id, threshold=args.threshold, threshold_col_id=args.threshold_col_id)\n",
    "    CDS_start_points1 = get_features(gff=args.gff, gene_id=args.gene_id, feature=args.feature, search_gff=args.search_gff, strand=args.strand, attribute=args.attribute)\n",
    "    create_background_fasta(CDS_start_points=CDS_start_points1, gene_id=args.gene_id, genome=args.genome, background_outfile=args.background_outfile, upstream_nucl=args.upstream_nucl)\n",
    "    create_target_fasta(DEgenes=DEgenes1, CDS_start_points=CDS_start_points1, gene_id=args.gene_id, genome=args.genome, target_outfile=args.target_outfile, upstream_nucl=args.upstream_nucl)\n",
    "\n",
    "    \n",
    "if __name__==\"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
